botorch/profiling.py

```diff
@@ -0,0 +1,247 @@
+    """
+    profile_MCMC_run = False
+    profile_fit_fully_bayesian_model_nuts = False
+
+    train_X, train_Y = _get_train_x_and_y(n, dim)
+
+    gp = SaasFullyBayesianSingleTaskGP(train_X=train_X, train_Y=train_Y)
+    # replace existing decorator with time_profile
+    import pyro
+
+    if profile_MCMC_run:
+        with mock.patch("pyro.poutine.block", time_profile):
+            import importlib
+
+            importlib.reload(pyro.infer.mcmc.api)
+
+    from pyro.infer.mcmc.api import MCMC
+
+    with (
+        mock.patch.object(pyro.infer.mcmc.MCMC, "run", MCMC.run),
+        mock.patch(
+            "pyro.infer.mcmc.api._UnarySampler.run",
+            time_profile(pyro.infer.mcmc.api._UnarySampler.run),
+        ),
+        mock.patch(
+            "pyro.infer.mcmc.api._MultiSampler.run",
+            time_profile(pyro.infer.mcmc.api._MultiSampler.run),
+        ),
+        mock.patch(
+            "pyro.infer.mcmc.api._gen_samples",
+            time_profile(pyro.infer.mcmc.api._gen_samples),
+        ),
+    ):
+        fit_fn = (
+            time_profile(fit_fully_bayesian_model_nuts)
+            if profile_fit_fully_bayesian_model_nuts
+            else fit_fully_bayesian_model_nuts
+        )
+        fit_fn(gp, warmup_steps=32, num_samples=16, thinning=16, disable_progbar=True)
+
+
+@memory_profile
+def run_large_t_batch_posterior_sampling(
+    b: int = 1000000,
+    dim: int = 1,
+    n_train: int = 2,
+    train_batch_dim: int = 1,
+    n_points: int = 1,
+) -> None:
+    """
+    Similar to `run_qnei`, it's interesting that memory usage _stays_ elevated after
+    calling it. Due to either caching or an issue with garbage collection.
+
+        Line #    Mem usage    Increment  Occurrences   Line Contents
+    =============================================================
+    209    175.8 MiB    175.8 MiB           1   @memory_profile
+    210                                         def run_large_t_batch_posterior_sampling(  # noqa: E501
+    211                                             b: int = 1000000,
+    212                                             dim: int = 1,
+    213                                             n_train: int = 2,
+    214                                             train_batch_dim: int = 1,
+    215                                             n_points: int = 1,
+    216                                         ) -> None:
+    222    176.0 MiB      0.2 MiB           1       train_X, train_Y = _get_train_x_and_y(n_train, dim, train_batch_dim)
+    223
+    224    177.1 MiB      1.1 MiB           1       model = SingleTaskGP(train_X, train_Y)
+    225    178.9 MiB      1.8 MiB           1       posterior = model.posterior(train_X[0, :, :])
+    226                                             # memory_profile(posterior.sample)(sample_shape=torch.Size([b, n_points]))
+    227    224.9 MiB     46.0 MiB           1       posterior.sample(sample_shape=torch.Size([b, n_points]))
```

Comment:
this here is explained by the lazy evaluation of the covariance in the MVN that is contained in the `Posterior` object generated by the `posterior()` call.  `model.posterior` will generate that `Posterior` which has an `MVN` whose covariance matrix is implicitly represented as a `LinearOperator`. Upon the first `posterior.sample` call, that covariance matrix is actually instantiated and decomposed via a Cholesky decomposition. This decomposition is cached. Since all subsequent `sample` calls just do a matrix-vector multiplication with that cached Cholesky decomposition on the same `Posterior` and hence `MVN` object, this doesn't really need to allocate memory and is also very fast.